cosmos/cosmos-sdk
# feat: explicit config file entry for ABCI client type

# Description

Refs: Agoric/agoric-sdk#9224

Add an explicit entry for `abci-client-type` in the TOML file. The setting can be done in the file but is overridden by when the command-line flag is set.

Note that this entry is unlike the other entries in the "base config" part of the TOML file, in that there are no `BaseApp.Get/SetABCIClientType()` methods, option helpers, etc. This makes it more like the various start command flags that do not have explicit config fields, e.g. `--with-tendermint` or `--trace-store`. This seems appropriate for a setting that is used near the "top" of the start command.

Note that even without this change, the option can be set in the `config/app.toml` file. This just makes the entry part of the automatically-generated TOML file.

Note that the default value is now specified both in the flag definition and in the default config constructor, as some of the other flags are. It's confusing to think about which one takes precedence, so they should just be kept in sync.

Tested the following:

- start with default toml file which includes and explicit entry to the default "committing" value
- start with an illegal value - fails as expected
- start with non-default valid value "local"
- start with line removed from `app.toml`
- override with flag set to an illegal value - fails as expected
- override with flag set to "committing"
- restore `app.toml` entry and override with a different legal value
- override with illegal flag - still fails as expected.

---

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage
